transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,698 @@
+                        };
+                        Poll::Ready(Some(event))
+                    }
+                    UDPMuxEvent::Error(e) => {
+                        self.close(Err(Error::UDPMuxError(e)));
+                        return self.poll_next(cx);
+                    }
+                    _ => return self.poll_next(cx),
+                }
+            }
+        }
+    }
+}
+
+// TODO: remove
+fn hex_to_cow<'a>(s: &str) -> Cow<'a, [u8; 32]> {
+    let mut buf = [0; 32];
+    hex::decode_to_slice(s, &mut buf).unwrap();
+    Cow::Owned(buf)
+}
+
+/// Turns an IP address and port into the corresponding WebRTC multiaddr.
+pub(crate) fn socketaddr_to_multiaddr(socket_addr: &SocketAddr) -> Multiaddr {
+    // TODO: remove
+    let f = "ACD1E533EC271FCDE0275947F4D62A2B2331FF10C9DDE0298EB7B399B4BFF60B";
+    Multiaddr::empty()
+        .with(socket_addr.ip().into())
+        .with(Protocol::Udp(socket_addr.port()))
+        .with(Protocol::XWebRTC(hex_to_cow(&f)))
+}
+
+/// Extracts a SHA-256 fingerprint from the given address. Returns `None` if the address does not
+/// contain one.
+fn fingerprint_from_addr<'a>(addr: &'a Multiaddr) -> Option<Cow<'a, [u8; 32]>> {
+    let iter = addr.iter();
+    for proto in iter {
+        match proto {
+            // TODO: check hash is one of https://datatracker.ietf.org/doc/html/rfc8122#section-5
+            Protocol::XWebRTC(f) => return Some(f),
+            _ => continue,
+        }
+    }
+    None
+}
+
+/// Transforms a byte array fingerprint into a string.
+pub(crate) fn fingerprint_to_string<T>(f: T) -> String
+where
+    T: IntoIterator,
+    <T as IntoIterator>::Item: core::fmt::LowerHex,
+{
+    let values: Vec<String> = f.into_iter().map(|x| format! {"{:02x}", x}).collect();
+    values.join(":")
+}
+
+/// Tries to turn a WebRTC multiaddress into a [`SocketAddr`]. Returns None if the format of the
+/// multiaddr is wrong.
+fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Option<SocketAddr> {
+    let mut iter = addr.iter();
+    let proto1 = iter.next()?;
+    let proto2 = iter.next()?;
+    let proto3 = iter.next()?;
+
+    for proto in iter {
+        match proto {
+            Protocol::P2p(_) => {} // Ignore a `/p2p/...` prefix of possibly outer protocols, if present.
+            _ => return None,
+        }
+    }
+
+    match (proto1, proto2, proto3) {
+        (Protocol::Ip4(ip), Protocol::Udp(port), Protocol::XWebRTC(_)) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        (Protocol::Ip6(ip), Protocol::Udp(port), Protocol::XWebRTC(_)) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        _ => None,
+    }
+}
+
+/// Returns a SHA-256 fingerprint of the first certificate.
+///
+/// # Panics
+///
+/// Panics if the config does not contain any certificates.
+pub(crate) fn fingerprint_of_first_certificate(config: &RTCConfiguration) -> String {
+    // safe to unwrap here because we require a certificate during construction.
```

Comment:
Such invariants could be enforced with the type system if we create our own config type.